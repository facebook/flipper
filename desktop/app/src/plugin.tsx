/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import {KeyboardActions} from './MenuBar';
import {Logger} from './fb-interfaces/Logger';
import Client from './Client';
import {Component} from 'react';
import BaseDevice from './devices/BaseDevice';
import {StaticView} from './reducers/connections';
import {State as ReduxState} from './reducers';
import {DEFAULT_MAX_QUEUE_SIZE} from './reducers/pluginMessageQueue';
import {ActivatablePluginDetails} from 'flipper-plugin-lib';
import {Settings} from './reducers/settings';
import {
  Notification,
  Idler,
  _SandyPluginDefinition,
  _makeShallowSerializable,
  _deserializeShallowObject,
} from 'flipper-plugin';

type Parameters = {[key: string]: any};

export type PluginDefinition = _SandyPluginDefinition;

export type ClientPluginMap = Map<string, PluginDefinition>;
export type DevicePluginMap = Map<string, PluginDefinition>;

// This function is intended to be called from outside of the plugin.
// If you want to `call` from the plugin use, this.client.call
export function callClient(
  client: Client,
  id: string,
): (method: string, params: Parameters) => Promise<any> {
  return (method, params) => client.call(id, method, false, params);
}

// This function is intended to be called from outside of the plugin.
// If you want to `supportsMethod` from the plugin use, this.client.supportsMethod
export function supportsMethod(
  client: Client,
  id: string,
): (method: string) => Promise<boolean> {
  return (method) => client.supportsMethod(id, method);
}

export interface PluginClient {
  isConnected: boolean;
  // eslint-disable-next-line
  send(method: string, params?: Parameters): void;
  // eslint-disable-next-line
  call(method: string, params?: Parameters): Promise<any>;
  // eslint-disable-next-line
  subscribe(method: string, callback: (params: any) => void): void;
  // eslint-disable-next-line
  supportsMethod(method: string): Promise<boolean>;
}

type PluginTarget = BaseDevice | Client;

export type Props<T> = {
  logger: Logger;
  persistedState: T;
  setPersistedState: (state: Partial<T>) => void;
  target: PluginTarget;
  deepLinkPayload: unknown;
  selectPlugin: (pluginID: string, deepLinkPayload: unknown) => void;
  isArchivedDevice: boolean;
  selectedApp: string | null;
  setStaticView: (payload: StaticView) => void;
  settingsState: Settings;
};

export type BaseAction = {
  type: string;
};

export type PersistedStateReducer = (
  persistedState: StaticPersistedState,
  method: string,
  data: any,
) => StaticPersistedState;

type StaticPersistedState = any;

export abstract class FlipperBasePlugin<
  State,
  Actions extends BaseAction,
  PersistedState,
> extends Component<Props<PersistedState>, State> {
  abstract ['constructor']: any;
  static title: string | null = null;
  static category: string | null = null;
  static id: string = '';
  static packageName: string = '';
  static version: string = '';
  static icon: string | null = null;
  static gatekeeper: string | null = null;
  static isBundled: boolean;
  static details: ActivatablePluginDetails;
  static keyboardActions: KeyboardActions | null;
  static screenshot: string | null;
  static defaultPersistedState: any;
  static persistedStateReducer: PersistedStateReducer | null;
  static maxQueueSize: number = DEFAULT_MAX_QUEUE_SIZE;
  static exportPersistedState:
    | ((
        callClient:
          | undefined
          | ((method: string, params?: any) => Promise<any>),
        persistedState: StaticPersistedState | undefined,
        store: ReduxState | undefined,
        idler?: Idler,
        statusUpdate?: (msg: string) => void,
        supportsMethod?: (method: string) => Promise<boolean>,
      ) => Promise<StaticPersistedState | undefined>)
    | undefined;
  static getActiveNotifications:
    | ((persistedState: StaticPersistedState) => Array<Notification>)
    | undefined;

  reducers: {
    [actionName: string]: (state: State, actionData: any) => Partial<State>;
  } = {};
  onKeyboardAction: ((action: string) => void) | undefined;

  toJSON() {
    return `<${this.constructor.name}#${this.constructor.id}>`;
  }

  // methods to be overriden by plugins
  init(): void {}

  static serializePersistedState: (
    persistedState: StaticPersistedState,
    statusUpdate?: (msg: string) => void,
    idler?: Idler,
    pluginName?: string,
  ) => Promise<string> = async (
    persistedState: StaticPersistedState,
    _statusUpdate?: (msg: string) => void,
    _idler?: Idler,
    _pluginName?: string,
  ) => {
    if (
      persistedState &&
      typeof persistedState === 'object' &&
      !Array.isArray(persistedState)
    ) {
      return JSON.stringify(
        Object.fromEntries(
          Object.entries(persistedState).map(([key, value]) => [
            key,
            _makeShallowSerializable(value), // make first level of persisted state serializable
          ]),
        ),
      );
    } else {
      return JSON.stringify(persistedState);
    }
  };

  static deserializePersistedState: (
    serializedString: string,
  ) => StaticPersistedState = (serializedString: string) => {
    const raw = JSON.parse(serializedString);
    if (raw && typeof raw === 'object' && !Array.isArray(raw)) {
      return Object.fromEntries(
        Object.entries(raw).map(([key, value]) => [
          key,
          _deserializeShallowObject(value),
        ]),
      );
    } else {
      return raw;
    }
  };

  teardown(): void {}

  // methods to be overridden by subclasses
  _init(): void {}

  _teardown(): void {}

  dispatchAction(actionData: Actions) {
    const action = this.reducers[actionData.type];
    if (!action) {
      throw new ReferenceError(`Unknown action ${actionData.type}`);
    }

    if (typeof action === 'function') {
      this.setState(action.call(this, this.state, actionData) as State);
    } else {
      throw new TypeError(`Reducer ${actionData.type} isn't a function`);
    }
  }
}

/**
 * @deprecated Please use the newer "Sandy" plugin APIs!
 * https://fbflipper.com/docs/extending/sandy-migration
 */
export class FlipperDevicePlugin<
  S,
  A extends BaseAction,
  P,
> extends FlipperBasePlugin<S, A, P> {
  ['constructor']: typeof FlipperPlugin;
  device: BaseDevice;

  constructor(props: Props<P>) {
    super(props);
    this.device = props.target as BaseDevice;
  }

  _init() {
    this.init();
  }

  _teardown() {
    this.teardown();
  }

  // TODO T84453692: remove this function after some transition period in favor of BaseDevice.supportsPlugin.
  static supportsDevice(_device: BaseDevice): boolean {
    throw new Error(
      'supportsDevice is unimplemented in FlipperDevicePlugin class',
    );
  }
}

/**
 * @deprecated Please use the newer "Sandy" plugin APIs!
 * https://fbflipper.com/docs/extending/sandy-migration
 */
export class FlipperPlugin<
  S,
  A extends BaseAction,
  P,
> extends FlipperBasePlugin<S, A, P> {
  ['constructor']: typeof FlipperPlugin;
  constructor(props: Props<P>) {
    super(props);
    // @ts-ignore constructor should be assigned already
    const {id} = this.constructor;
    this.subscriptions = [];
    const realClient = (this.realClient = props.target as Client);
    this.client = {
      get isConnected() {
        return realClient.connected.get();
      },
      call: (method, params) => this.realClient.call(id, method, true, params),
      send: (method, params) => this.realClient.send(id, method, params),
      subscribe: (method, callback) => {
        this.subscriptions.push({
          method,
          callback,
        });
        this.realClient.subscribe(id, method, callback);
      },
      supportsMethod: (method) => this.realClient.supportsMethod(id, method),
    };
  }

  subscriptions: Array<{
    method: string;
    callback: Function;
  }>;

  client: PluginClient;
  realClient: Client;

  /**
   * @deprecated use .device instead
   */
  getDevice(): Promise<BaseDevice> {
    return this.realClient.device;
  }

  get device() {
    return this.realClient.deviceSync;
  }

  _teardown() {
    // automatically unsubscribe subscriptions
    const pluginId = this.constructor.id;
    for (const {method, callback} of this.subscriptions) {
      this.realClient.unsubscribe(pluginId, method, callback);
    }
    // run plugin teardown
    this.teardown();
    if (!this.realClient.isBackgroundPlugin(pluginId)) {
      this.realClient.deinitPlugin(pluginId);
    }
  }

  _init() {
    const pluginId = this.constructor.id;
    if (!this.realClient.isBackgroundPlugin(pluginId)) {
      this.realClient.initPlugin(pluginId);
    }
    this.init();
  }
}
