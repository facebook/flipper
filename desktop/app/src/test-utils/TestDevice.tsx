/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import {DeviceOS, DeviceType} from 'flipper-plugin';
import {DeviceSpec} from 'flipper-plugin-lib';
import BaseDevice from '../devices/BaseDevice';

export class TestDevice extends BaseDevice {
  constructor(
    serial: string,
    deviceType: DeviceType,
    title: string,
    os: DeviceOS,
    specs?: DeviceSpec[],
  ) {
    super(
      {
        on: jest.fn(),
        off: jest.fn(),
        exec: jest.fn(),
        close: jest.fn(),
      },
      {
        serial,
        deviceType,
        title,
        os,
        specs,
      },
    );
  }
}
