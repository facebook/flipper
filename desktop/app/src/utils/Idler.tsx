/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import {CancelledPromiseError} from './errors';
import {Idler, sleep} from 'flipper-plugin';

export class IdlerImpl implements Idler {
  private lastIdle = performance.now();
  private kill = false;

  constructor(private interval = 16) {}

  shouldIdle(): boolean {
    return this.kill || performance.now() - this.lastIdle > this.interval;
  }

  async idle(): Promise<void> {
    if (this.kill) {
      throw new CancelledPromiseError('Idler got killed');
    }
    const now = performance.now();
    if (now - this.lastIdle > this.interval) {
      this.lastIdle = now;
      return new Promise((resolve) => {
        if (typeof requestIdleCallback !== 'undefined') {
          requestIdleCallback(() => {
            resolve();
          });
        } else {
          setTimeout(resolve, 0);
        }
      });
    }
    return undefined;
  }

  cancel() {
    this.kill = true;
  }

  isCancelled() {
    return this.kill;
  }
}

// This smills like we should be using generators :)
export class TestIdler implements Idler {
  private resolver?: () => void;
  private kill = false;
  private autoRun = false;
  private hasProgressed = false;

  constructor(autorun = false) {
    this.autoRun = autorun;
  }

  shouldIdle() {
    if (this.kill) {
      return true;
    }
    if (this.autoRun) {
      return false;
    }
    // In turns we signal that idling is needed and that it isn't
    if (!this.hasProgressed) {
      this.hasProgressed = true;
      return false;
    }
    return true;
  }

  async idle() {
    if (this.kill) {
      throw new CancelledPromiseError('Idler got killed');
    }
    if (this.autoRun) {
      return undefined;
    }
    if (this.resolver) {
      throw new Error('Already idling');
    }
    return new Promise<void>((resolve) => {
      this.resolver = () => {
        this.resolver = undefined;
        this.hasProgressed = false;
        resolve();
      };
    });
  }

  cancel() {
    this.kill = true;
    this.run();
  }

  async next() {
    if (!this.resolver) {
      throw new Error('Not yet idled');
    }

    this.resolver();
    // make sure waiting promise runs first
    await sleep(10);
  }

  /**
   * Automatically progresses through all idle calls
   */
  run() {
    this.resolver?.();
    this.autoRun = true;
  }

  isCancelled() {
    return this.kill;
  }
}
