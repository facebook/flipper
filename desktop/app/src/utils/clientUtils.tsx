/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import type Client from '../Client';
import type BaseDevice from '../devices/BaseDevice';

/* A Client uniuely identifies an app running on some device.

  Always use this utility to construct and parse clientId strings.
 */
export type ClientIdConstituents = {
  app: string;
  os: string;
  device: string;
  device_id: string;
};

/* A plugin key is a string uniquely identifying an instance of a plugin.
   This can be a device plugin for a particular device, or a client plugin for a particular client (app).
   In the device plugin case, the "client" is the device it's connected to.
   In the client plugin case (normal plugins), the "client" is the app it's connected to.

   Always use this utility to construct and parse pluginKey strings.
 */
type PluginKeyConstituents =
  | {
      type: 'device';
      pluginName: string;
      client: string;
    }
  | ({
      type: 'client';
      pluginName: string;
      client: string;
    } & ClientIdConstituents);

export function currentActiveApps(
  clients: Array<Client>,
  selectedDevice: null | BaseDevice,
): Array<string> {
  const currentActiveApps: Array<string> = clients
    .map(({id}: {id: string}) => {
      const appName = deconstructClientId(id).app || '';
      const os = deconstructClientId(id).os || '';
      return {appName, os};
    })
    .filter(
      ({os}: {os: string}) => os && selectedDevice && os == selectedDevice.os,
    )
    .map((client) => client.appName);
  return currentActiveApps;
}

export function buildClientId(clientInfo: {
  app: string;
  os: string;
  device: string;
  device_id: string;
}): string {
  // N.B.: device_id can be empty, which designates the host device
  for (const key of ['app', 'os', 'device'] as Array<
    keyof ClientIdConstituents
  >) {
    if (!clientInfo[key]) {
      console.error(
        `Attempted to build clientId with invalid ${key}: "${clientInfo[key]}`,
      );
    }
  }
  const escapedName = escape(clientInfo.app);
  return `${escapedName}#${clientInfo.os}#${clientInfo.device}#${clientInfo.device_id}`;
}

export function deconstructClientId(clientId: string): ClientIdConstituents {
  if (!clientId || clientId.split('#').length !== 4) {
    console.error(`Attempted to deconstruct invalid clientId: "${clientId}"`);
  }
  let [app, os, device, device_id] = clientId.split('#');
  app = unescape(app);
  return {
    app,
    os,
    device,
    device_id,
  };
}

export function deconstructPluginKey(pluginKey: string): PluginKeyConstituents {
  const parts = pluginKey.split('#');
  if (parts.length === 2) {
    // Device plugin
    return {
      type: 'device',
      client: parts[0],
      pluginName: parts[1],
    };
  } else {
    // Client plugin
    const lastHashIndex = pluginKey.lastIndexOf('#');
    const clientId = pluginKey.slice(0, lastHashIndex);
    const pluginName = pluginKey.slice(lastHashIndex + 1);
    if (!pluginName) {
      console.error(
        `Attempted to deconstruct invalid pluginKey: "${pluginKey}"`,
      );
    }
    return {
      type: 'client',
      ...deconstructClientId(clientId),
      client: clientId,
      pluginName: pluginName,
    };
  }
}
