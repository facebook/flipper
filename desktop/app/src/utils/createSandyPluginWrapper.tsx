/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import * as React from 'react';
import {
  createState,
  useLogger,
  usePlugin,
  useValue,
  _SandyPluginDefinition,
  PluginClient,
  DevicePluginClient,
} from 'flipper-plugin';
import {useEffect} from 'react';
import {
  BaseAction,
  FlipperDevicePlugin,
  FlipperPlugin,
  Props as PluginProps,
} from '../plugin';
import {useStore} from './useStore';
import {setStaticView, StaticView} from '../reducers/connections';
import {getStore} from '../store';
import {setActiveNotifications} from '../reducers/notifications';

export type SandyPluginModule = ConstructorParameters<
  typeof _SandyPluginDefinition
>[1];

export function createSandyPluginWrapper<S, A extends BaseAction, P>(
  Plugin: typeof FlipperPlugin | typeof FlipperDevicePlugin,
): SandyPluginModule {
  const isDevicePlugin = Plugin.prototype instanceof FlipperDevicePlugin;
  if (process.env.NODE_ENV !== 'test') {
    console.warn(
      `Loading ${isDevicePlugin ? 'device' : 'client'} plugin ${
        Plugin.id
      } in legacy mode. Please visit https://fbflipper.com/docs/extending/sandy-migration to learn how to migrate this plugin to the new Sandy architecture`,
    );
  }

  function legacyPluginWrapper(client: PluginClient | DevicePluginClient) {
    const store = getStore();
    const appClient = isDevicePlugin
      ? undefined
      : (client as PluginClient<any, any>);

    const instanceRef = React.createRef<FlipperPlugin<S, A, P> | null>();
    const persistedState = createState<P>(Plugin.defaultPersistedState);
    const deeplink = createState<unknown>();

    client.onDeepLink((link) => {
      deeplink.set(link);
    });

    appClient?.onUnhandledMessage((event, params) => {
      if (Plugin.persistedStateReducer) {
        persistedState.set(
          Plugin.persistedStateReducer(persistedState.get(), event, params),
        );
      }
    });

    if (
      Plugin.persistedStateReducer ||
      Plugin.exportPersistedState ||
      Plugin.defaultPersistedState
    ) {
      client.onExport(async (idler, onStatusMessage) => {
        const state = Plugin.exportPersistedState
          ? await Plugin.exportPersistedState(
              isDevicePlugin
                ? undefined
                : (method: string, params: any) =>
                    appClient!.send(method, params),
              persistedState.get(),
              undefined, // passing an undefined Store is safe, as no plugin actually uses this param
              idler,
              onStatusMessage,
              isDevicePlugin
                ? undefined
                : (method: string) => appClient!.supportsMethod(method),
            )
          : persistedState.get();
        // respect custom serialization
        return Plugin.serializePersistedState
          ? await Plugin.serializePersistedState(
              state,
              onStatusMessage,
              idler,
              Plugin.id,
            )
          : state;
      });

      client.onImport((data) => {
        if (Plugin.deserializePersistedState) {
          data = Plugin.deserializePersistedState(data);
        }
        persistedState.set(data);
      });
    }

    if (Plugin.keyboardActions) {
      function executeKeyboardAction(action: string) {
        instanceRef?.current?.onKeyboardAction?.(action);
      }

      client.addMenuEntry(
        ...Plugin.keyboardActions.map((def) => {
          if (typeof def === 'string') {
            return {
              action: def,
              handler() {
                executeKeyboardAction(def);
              },
            };
          } else {
            const {action, label, accelerator, topLevelMenu} = def;
            return {
              label,
              accelerator,
              topLevelMenu,
              handler() {
                executeKeyboardAction(action);
              },
            };
          }
        }),
      );
    }

    if (Plugin.getActiveNotifications && !isDevicePlugin) {
      const unsub = persistedState.subscribe((state) => {
        try {
          const notifications = Plugin.getActiveNotifications!(state);
          store.dispatch(
            setActiveNotifications({
              notifications,
              client: appClient!.appId,
              pluginId: Plugin.id,
            }),
          );
        } catch (e) {
          console.error(
            'Failed to compute notifications for plugin ' + Plugin.id,
            e,
          );
        }
      });
      client.onDestroy(unsub);
    }

    return {
      instanceRef,
      device: client.device.realDevice,
      persistedState,
      deeplink,
      selectPlugin: client.selectPlugin,
      setPersistedState(state: Partial<P>) {
        persistedState.set({...persistedState.get(), ...state});
      },
      get appId() {
        return appClient?.appId;
      },
      get appName() {
        return appClient?.appName ?? null;
      },
      get isArchived() {
        return client.device.isArchived;
      },
      setStaticView(payload: StaticView) {
        store.dispatch(setStaticView(payload));
      },
    };
  }

  function Component() {
    const instance = usePlugin(legacyPluginWrapper);
    const logger = useLogger();
    const persistedState = useValue(instance.persistedState);
    const deepLinkPayload = useValue(instance.deeplink);
    const settingsState = useStore((state) => state.settingsState);

    const target = isDevicePlugin
      ? instance.device
      : // eslint-disable-next-line
          useStore((state) =>
          state.connections.clients.find((c) => c.id === instance.appId),
        );
    if (!target) {
      throw new Error('Illegal state: missing target');
    }

    useEffect(
      function triggerInitAndTeardown() {
        const ref = instance.instanceRef.current!;
        ref._init();
        return () => {
          ref._teardown();
        };
      },
      [instance.instanceRef],
    );

    const props: PluginProps<P> = {
      logger,
      persistedState,
      deepLinkPayload,
      settingsState,
      target,
      setPersistedState: instance.setPersistedState,
      selectPlugin: instance.selectPlugin,
      isArchivedDevice: instance.isArchived,
      selectedApp: instance.appName,
      setStaticView: instance.setStaticView,
      // @ts-ignore ref is not on Props
      ref: instance.instanceRef,
    };

    return React.createElement(Plugin, props);
  }

  return isDevicePlugin
    ? {devicePlugin: legacyPluginWrapper, Component}
    : {
        plugin: legacyPluginWrapper,
        Component,
      };
}
