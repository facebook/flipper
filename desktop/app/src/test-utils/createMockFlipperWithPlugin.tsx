/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import React from 'react';
import {Provider} from 'react-redux';
import {
  render,
  RenderResult,
  act as testingLibAct,
} from '@testing-library/react';
import {queries} from '@testing-library/dom';

import {
  selectPlugin,
  selectDevice,
  selectClient,
} from '../reducers/connections';
import BaseDevice from '../devices/BaseDevice';

import {Store} from '../reducers/index';
import Client, {ClientQuery} from '../Client';

import {Logger} from '../fb-interfaces/Logger';
import {FlipperDevicePlugin, FlipperPlugin, PluginDefinition} from '../plugin';
import PluginContainer from '../PluginContainer';
import {getPluginKey, isDevicePluginDefinition} from '../utils/pluginUtils';
import MockFlipper from './MockFlipper';
import {switchPlugin} from '../reducers/pluginManager';
import {createSandyPluginFromClassicPlugin} from '../dispatcher/plugins';
import {createMockActivatablePluginDetails} from '../utils/testUtils';
import {_SandyPluginDefinition} from 'flipper-plugin';

export type MockFlipperResult = {
  client: Client;
  device: BaseDevice;
  store: Store;
  pluginKey: string;
  sendError(error: any, client?: Client): void;
  sendMessage(method: string, params: any, client?: Client): void;
  createDevice(serial: string): BaseDevice;
  createClient(
    device: BaseDevice,
    name: string,
    query?: ClientQuery,
    skipRegister?: boolean,
  ): Promise<Client>;
  logger: Logger;
  togglePlugin(plugin?: string): void;
  selectPlugin(
    id?: string,
    client?: Client,
    device?: BaseDevice,
    deepLinkPayload?: any,
  ): void;
};

type MockOptions = Partial<{
  /**
   * can be used to intercept outgoing calls. If it returns undefined
   * the base implementation will be used
   */
  onSend?: (pluginId: string, method: string, params?: object) => any;
  additionalPlugins?: (PluginDefinition | LegacyPluginDefinition)[];
  dontEnableAdditionalPlugins?: true;
  asBackgroundPlugin?: true;
  supportedPlugins?: string[];
  device?: BaseDevice;
  archivedDevice?: boolean;
}>;

function isPluginEnabled(
  store: Store,
  pluginClazz: PluginDefinition,
  selectedApp: string,
) {
  return (
    (!isDevicePluginDefinition(pluginClazz) &&
      store
        .getState()
        .connections.enabledPlugins[selectedApp]?.includes(pluginClazz.id)) ||
    (isDevicePluginDefinition(pluginClazz) &&
      store.getState().connections.enabledDevicePlugins.has(pluginClazz.id))
  );
}

export type LegacyPluginDefinition =
  | typeof FlipperDevicePlugin
  | typeof FlipperPlugin;

export function wrapSandy(
  clazz: PluginDefinition | LegacyPluginDefinition,
): PluginDefinition {
  return clazz instanceof _SandyPluginDefinition
    ? clazz
    : createSandyPluginFromClassicPlugin(
        createMockActivatablePluginDetails({
          id: clazz.id,
          title: clazz.title ?? clazz.id,
          pluginType:
            clazz.prototype instanceof FlipperDevicePlugin
              ? 'device'
              : 'client',
        }),
        clazz,
      );
}

export async function createMockFlipperWithPlugin(
  pluginClazzOrig: PluginDefinition | LegacyPluginDefinition,
  options?: MockOptions,
): Promise<MockFlipperResult> {
  const pluginClazz = wrapSandy(pluginClazzOrig);
  const additionalPlugins = options?.additionalPlugins?.map(wrapSandy) ?? [];

  const mockFlipper = new MockFlipper();
  await mockFlipper.init({
    plugins: [pluginClazz, ...additionalPlugins],
  });
  const logger = mockFlipper.logger;
  const store = mockFlipper.store;

  const createDevice = (serial: string, archived?: boolean) =>
    mockFlipper.createDevice({serial, archived});
  const createClient = async (
    device: BaseDevice,
    name: string,
    query?: ClientQuery,
    skipRegister?: boolean,
  ) => {
    const client = await mockFlipper.createClient(device, {
      name,
      query,
      skipRegister,
      onSend: options?.onSend,
      supportedPlugins: options?.supportedPlugins,
      backgroundPlugins: options?.asBackgroundPlugin ? [pluginClazz.id] : [],
    });

    // enable the plugin
    if (!isPluginEnabled(store, pluginClazz, name)) {
      store.dispatch(
        switchPlugin({
          plugin: pluginClazz,
          selectedApp: client.query.app,
        }),
      );
    }

    if (!options?.dontEnableAdditionalPlugins) {
      additionalPlugins.forEach((plugin) => {
        if (!isPluginEnabled(store, plugin, name)) {
          store.dispatch(
            switchPlugin({
              plugin,
              selectedApp: client.query.app,
            }),
          );
        }
      });
    }
    return client;
  };

  const device = options?.device
    ? mockFlipper.loadDevice(options?.device)
    : createDevice('serial', options?.archivedDevice);
  const client = await createClient(device, 'TestApp');

  store.dispatch(selectDevice(device));
  store.dispatch(selectClient(client.id));

  let lastSelected: string | undefined = undefined;

  function selectPluginImpl(
    id = pluginClazz.id,
    theClient = client,
    theDevice = device,
    deepLinkPayload = null,
  ) {
    if (lastSelected) {
      client.deinitPlugin(lastSelected);
    }
    store.dispatch(
      selectPlugin({
        selectedPlugin: id,
        selectedApp: theClient.query.app,
        deepLinkPayload,
        selectedDevice: theDevice,
      }),
    );
    client.initPlugin(pluginClazz.id); // simulates plugin being mounted
    lastSelected = pluginClazz.id;
  }

  selectPluginImpl();

  return {
    client,
    device: device as any,
    store,
    selectPlugin: selectPluginImpl,
    sendError(error: any, actualClient = client) {
      actualClient.onMessage(
        JSON.stringify({
          error,
        }),
      );
    },
    sendMessage(method, params, actualClient = client) {
      actualClient.onMessage(
        JSON.stringify({
          method: 'execute',
          params: {
            api: pluginClazz.id,
            method,
            params,
          },
        }),
      );
    },
    createDevice,
    createClient,
    logger,
    pluginKey: getPluginKey(client.id, device, pluginClazz.id),
    togglePlugin(id?: string) {
      const plugin = id
        ? store.getState().plugins.clientPlugins.get(id) ??
          store.getState().plugins.devicePlugins.get(id)
        : pluginClazz;
      if (!plugin) {
        throw new Error('unknown plugin ' + id);
      }
      store.dispatch(
        switchPlugin({
          plugin,
          selectedApp: client.query.app,
        }),
      );
    },
  };
}

type Renderer = RenderResult<typeof queries>;

export async function renderMockFlipperWithPlugin(
  pluginClazzOrig: PluginDefinition | LegacyPluginDefinition,
  options?: MockOptions,
): Promise<
  MockFlipperResult & {
    renderer: Renderer;
    act: (cb: () => void) => void;
  }
> {
  const pluginClazz = wrapSandy(pluginClazzOrig);
  const args = await createMockFlipperWithPlugin(pluginClazz, options);

  function selectTestPlugin(store: Store, client: Client) {
    store.dispatch(
      selectPlugin(
        isDevicePluginDefinition(pluginClazz)
          ? {
              selectedPlugin: pluginClazz.id,
              selectedApp: null,
              deepLinkPayload: null,
              selectedDevice: store.getState().connections.selectedDevice!,
            }
          : {
              selectedPlugin: pluginClazz.id,
              selectedApp: client.query.app,
              deepLinkPayload: null,
              selectedDevice: store.getState().connections.selectedDevice!,
            },
      ),
    );
  }

  selectTestPlugin(args.store, args.client);

  const renderer = render(
    <Provider store={args.store}>
      <PluginContainer logger={args.logger} />
    </Provider>,
  );

  return {
    ...args,
    renderer: renderer as any,
    act(cb) {
      testingLibAct(cb);
      args.client.flushMessageBuffer();
    },
  };
}
