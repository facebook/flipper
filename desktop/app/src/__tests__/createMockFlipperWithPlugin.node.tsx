/**
 * Copyright (c) Facebook, Inc. and its affiliates.
 *
 * This source code is licensed under the MIT license found in the
 * LICENSE file in the root directory of this source tree.
 *
 * @format
 */

import {createMockFlipperWithPlugin} from '../test-utils/createMockFlipperWithPlugin';
import {FlipperPlugin} from '../plugin';
import {TestIdler} from '../utils/Idler';

interface PersistedState {
  count: 1;
}

class TestPlugin extends FlipperPlugin<any, any, any> {
  static id = 'TestPlugin';

  static defaultPersistedState = {
    count: 0,
  };

  static persistedStateReducer(
    persistedState: PersistedState,
    method: string,
    _payload: {},
  ) {
    if (method === 'inc') {
      return Object.assign({}, persistedState, {
        count: persistedState.count + 1,
      });
    }
    return persistedState;
  }

  render() {
    return null;
  }
}

const testIdler = new TestIdler();

function testOnStatusMessage() {
  // emtpy stub
}

test('can create a Fake flipper with legacy wrapper', async () => {
  const {client, device, store, sendMessage} =
    await createMockFlipperWithPlugin(TestPlugin);
  expect(client).toBeTruthy();
  expect(device).toBeTruthy();
  expect(store).toBeTruthy();
  expect(sendMessage).toBeTruthy();
  expect(client.plugins.has(TestPlugin.id)).toBe(true);
  expect(client.sandyPluginStates.has(TestPlugin.id)).toBe(true);
  const state = store.getState();
  expect(state.connections).toMatchSnapshot();
  expect(state.plugins).toMatchSnapshot();
  sendMessage('inc', {});
  expect(
    await state.connections.clients[0].sandyPluginStates
      .get(TestPlugin.id)!
      .exportState(testIdler, testOnStatusMessage),
  ).toMatchInlineSnapshot(`"{\\"count\\":1}"`);
});
