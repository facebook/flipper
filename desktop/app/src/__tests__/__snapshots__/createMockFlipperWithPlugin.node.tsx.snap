// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`can create a Fake flipper with legacy wrapper 1`] = `
Object {
  "androidEmulators": Array [],
  "clients": Array [
    Object {
      "id": "TestApp#Android#MockAndroidDevice#serial",
      "query": Object {
        "app": "TestApp",
        "device": "MockAndroidDevice",
        "device_id": "serial",
        "os": "Android",
        "sdk_version": 4,
      },
    },
  ],
  "deepLinkPayload": null,
  "devices": Array [
    Object {
      "deviceType": "physical",
      "os": "Android",
      "serial": "serial",
      "title": "MockAndroidDevice",
    },
  ],
  "enabledDevicePlugins": Set {
    "DeviceLogs",
    "CrashReporter",
    "MobileBuilds",
    "Hermesdebuggerrn",
    "React",
  },
  "enabledPlugins": Object {
    "TestApp": Array [
      "TestPlugin",
    ],
  },
  "selectedApp": "TestApp#Android#MockAndroidDevice#serial",
  "selectedAppPluginListRevision": 0,
  "selectedDevice": Object {
    "deviceType": "physical",
    "os": "Android",
    "serial": "serial",
    "title": "MockAndroidDevice",
  },
  "selectedPlugin": "TestPlugin",
  "staticView": null,
  "uninitializedClients": Array [],
  "userPreferredApp": "TestApp#Android#MockAndroidDevice#serial",
  "userPreferredDevice": "MockAndroidDevice",
  "userPreferredPlugin": "TestPlugin",
}
`;

exports[`can create a Fake flipper with legacy wrapper 2`] = `
Object {
  "bundledPlugins": Map {},
  "clientPlugins": Map {
    "TestPlugin" => SandyPluginDefinition {
      "details": Object {
        "dir": "/Users/mock/.flipper/thirdparty/flipper-plugin-sample1",
        "entry": "./test/index.js",
        "id": "TestPlugin",
        "isActivatable": true,
        "isBundled": false,
        "main": "dist/bundle.js",
        "name": "flipper-plugin-hello",
        "pluginType": "client",
        "source": "src/index.js",
        "specVersion": 2,
        "title": "TestPlugin",
        "version": "0.1.0",
      },
      "id": "TestPlugin",
      "isDevicePlugin": false,
      "module": Object {
        "Component": [Function],
        "plugin": [Function],
      },
    },
  },
  "devicePlugins": Map {},
  "disabledPlugins": Array [],
  "failedPlugins": Array [],
  "gatekeepedPlugins": Array [],
  "initialised": false,
  "installedPlugins": Map {},
  "loadedPlugins": Map {},
  "marketplacePlugins": Array [],
  "selectedPlugins": Array [],
  "uninstalledPluginNames": Set {},
}
`;
